sdk/identity/identity/src/client/identityClient.ts

```diff
@@ -110,29 +180,38 @@ export class IdentityClient extends ServiceClient {
     clientId?: string,
     getTokenOptions?: GetTokenOptions
   ): Promise<AccessToken | null> {
-    const queryParameters: any = {
-      resource: this.mapScopesToResource(scopes),
-      "api-version": MsiApiVersion
-    };
+    let authRequestOptions: RequestPrepareOptions;
+    const resource = this.mapScopesToResource(scopes);
+    let expiresInParser: ((requestBody: any) => number) | undefined;
 
-    if (clientId) {
-      queryParameters.client_id = clientId;
+    // Detect which type of environment we are running in
+    if (process.env.MSI_ENDPOINT) {
+      if (process.env.MSI_SECRET) {
+        // Running in App Service
+        authRequestOptions = this.createAppServiceMsiAuthRequest(resource, clientId);
+        expiresInParser = (requestBody: any) => {
+          // Parse a date format like "06/20/2019 02:57:58 +00:00" and
+          // convert it into a JavaScript-formatted date
+          const m = requestBody.expires_on.match(/(\d\d)\/(\d\d)\/(\d\d\d\d) (\d\d):(\d\d):(\d\d) (\+|-)(\d\d):(\d\d)/)
+          return Date.parse(`${m[3]}-${m[1]}-${m[2]}T${m[4]}:${m[5]}:${m[6]}${m[7]}${m[8]}:${m[9]}`)
+        };
+      } else {
+        // Running in Cloud Shell
+        authRequestOptions = this.createCloudShellMsiAuthRequest(resource, clientId);
+      }
+    } else {
+      // Running in an Azure VM
+      authRequestOptions = this.createAzureVmMsiAuthRequest(resource, clientId);
```

Comment:
If `MSI_ENDPOINT` and `MSI_SECRET` are undefined, does this just send a request to the IMDS endpoint? What's the timeout?

Reply:
Yep, defaults to IMDS endpoint.  I haven't configured a timeout and `core-http` defaults to having no timeout.  What timeout are you using in Python?

Reply:
It's more a question of what happens when IMDS isn't available. In that case I suppose you'd get your stack's default timeout, i.e. too long. In Python the credential [probes the endpoint](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/identity/azure-identity/azure/identity/_internal.py#L97) during the initial `get_token` call and if that probe times out, `get_token` fails without sending a request thereafter.

Reply:
That's a solid approach, I'll add something similar.

Reply:
We definitely need to add this, otherwise any credentials behind the ManagedIdentityCredential will have to wait for this request to timeout before they execute.  As of right now this is the last in the chain used by DefaultAzureCredential, but this will change.

Reply:
However, since ManagedIdentityCredential is going to be the last credential in the chain for preview one, we could create an issue for this and move it to preview 2 if necessary.